File: amlip_node_py/amlip_node_py/types/AmlipId.py
# Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""AML-IP AmlipId data type API specification."""

from amlip_swig import AmlipId as cpp_AmlipId
from amlip_swig import ID_SIZE


class AmlipId(cpp_AmlipId):
    """
    TODO
    """

    def __init__(self, *args):
        super().__init__(*args)

    def set_id(self, new_id: str) -> None:

        if new_id > ID_SIZE:
            raise Exception(f'Max number of chars for AMLIP id is {ID_SIZE}')
        else:
            return cpp_AmlipId.id(self, new_id)

    def get_id(self) -> str:
        return ''.join(cpp_AmlipId.id(self))

    def is_defined(self) -> bool:
        return cpp_AmlipId.is_defined(self)

    @staticmethod
    def new_unique_id() -> 'AmlipId':
        return cpp_AmlipId.new_unique_id()

    @staticmethod
    def undefined_id() -> 'AmlipId':
        return cpp_AmlipId.undefined_id()

    # def __str__(self) -> str:
    #     return self.get_id()

File: amlip_cpp/src/cpp/types/AmlipId.cpp
// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*!
 * @file AmlipId.cpp
 */

#include <memory>
#include <string>

#include <types/AmlipId.hpp>

namespace eprosima {
namespace amlip {
namespace types {

const std::string AmlipId::UNDEFINED_NAME_= "Undefined";

AmlipId::AmlipId(const std::string& name)
{
    data_ = std::make_shared<AmlipIdDataType>(name);
}

AmlipId::AmlipId(
        const AmlipId& x)
{
    data_ = std::make_shared<AmlipIdDataType>(*x.data_);
}

AmlipId::AmlipId(
        AmlipId&& x)
{
    data_ = std::move(x.data_);
}

AmlipId::~AmlipId()
{
}

AmlipId& AmlipId::operator =(
    const AmlipId& x)
{
    data_ = x.data_;

    return *this;
}

AmlipId& AmlipId::operator =(
    AmlipId&& x)
{
    data_ = std::move(x.data_);

    return *this;
}

bool AmlipId::operator ==(
    const AmlipId& x) const
{
    return *data_ == *x.data_;
}

bool AmlipId::operator !=(
        const AmlipId& x) const
{
    return !(*this == x);
}

std::string AmlipId::name() const
{
    return data_->name();
}

std::shared_ptr<AmlipIdDataType> AmlipId::data() const
{
    return data_;
}

AmlipId AmlipId::new_unique_id(const std::string& name)
{
    return AmlipId(name);
}

std::ostream& operator <<(
        std::ostream& os,
        const AmlipId& id)
{
    return os << *id.data();
}

} /* namespace types */
} /* namespace amlip */
} /* namespace eprosima */
